mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala

```diff
@@ -0,0 +1,547 @@
+    val familyLink = $(family) match {
+      case "gaussian" => if (isDefined(link)) Gaussian($(link)) else Gaussian("identity")
+      case "binomial" => if (isDefined(link)) Binomial($(link)) else Binomial("logit")
+      case "poisson" => if (isDefined(link)) Poisson($(link)) else Poisson("log")
+      case "gamma" => if (isDefined(link)) Gamma($(link)) else Gamma("inverse")
+    }
+
+    val numFeatures = dataset.select(col($(featuresCol))).limit(1).map {
+      case Row(features: Vector) => features.size
+    }.first()
+    if (numFeatures > 4096) {
+      val msg = "Currently, GeneralizedLinearRegression only supports number of features" +
+        s" <= 4096. Found $numFeatures in the input dataset."
+      throw new SparkException(msg)
+    }
+
+    val w = if ($(weightCol).isEmpty) lit(1.0) else col($(weightCol))
+    val instances: RDD[Instance] = dataset.select(
+      col($(labelCol)), w, col($(featuresCol))).map {
+      case Row(label: Double, weight: Double, features: Vector) =>
+        Instance(label, weight, features)
+    }
+
+    if ($(family) == "gaussian" && $(link) == "identity") {
+      val optimizer = new WeightedLeastSquares($(fitIntercept), $(regParam),
+        standardizeFeatures = true, standardizeLabel = true)
+      val wlsModel = optimizer.fit(instances)
+      val model = copyValues(new GeneralizedLinearRegressionModel(uid,
+        wlsModel.coefficients, wlsModel.intercept).setParent(this))
+      return model
+    }
+
+    val newInstances = instances.map { instance =>
+      val mu = familyLink.initialize(instance.label, instance.weight)
+      val eta = familyLink.predict(mu)
+      Instance(eta, instance.weight, instance.features)
+    }
+
+    val initialModel = new WeightedLeastSquares($(fitIntercept), $(regParam),
+      standardizeFeatures = true, standardizeLabel = true).fit(newInstances)
+
+    val reweightFunc: (Instance, WeightedLeastSquaresModel) => (Double, Double) = {
+      (instance: Instance, model: WeightedLeastSquaresModel) => {
+        val eta = model.predict(instance.features)
+        val mu = familyLink.fitted(eta)
+        val z = familyLink.adjusted(instance.label, mu, eta)
+        val w = familyLink.weights(mu) * instance.weight
+        (z, w)
+      }
+    }
+
+    val optimizer = new IterativelyReweightedLeastSquares(initialModel, reweightFunc,
+      $(fitIntercept), $(regParam), $(maxIter), $(tol))
+
+    val irlsModel = optimizer.fit(instances)
+
+    val model = copyValues(new GeneralizedLinearRegressionModel(uid,
+      irlsModel.coefficients, irlsModel.intercept).setParent(this))
+    model
+  }
+
+  @Since("2.0.0")
+  override def copy(extra: ParamMap): GeneralizedLinearRegression = defaultCopy(extra)
+}
+
+@Since("2.0.0")
+object GeneralizedLinearRegression {
+
+  /** Set of families that GeneralizedLinearRegression supports */
+  private[ml] val supportedFamilies = Set("gaussian", "binomial", "poisson", "gamma")
```

Comment:
String references are error-prone. I would first define:

~~~scala
lazy val supportedFamilyAndLinkParis = Set(Gaussian(Identity), Gaussian(Log), ...)
~~~

and then 

~~~scala
lazy val supportedFamilyNames = supportedFamilies.map(_.name)
lazy val supportedLinkNames = supportedFamilies.map(_.link.name)
~~~